File: flutter_app/lib/core/application/use_cases/update_todo_use_case.dart
import '../../domain/entities/todo_entity.dart';
import '../../domain/entities/todo_icon_type.dart';
import '../../domain/repositories/todo_repository.dart';

class UpdateTodoUseCase {
  final TodoRepository _repository;

  UpdateTodoUseCase(this._repository);

  Future<TodoEntity> execute({
    required String id,
    String? title,
    String? description,
    DateTime? deadline,
    bool? isCompleted,
    TodoIconType? iconType,
  }) async {
    if (id.trim().isEmpty) {
      throw ArgumentError('ID cannot be empty');
    }

    if (title != null && title.trim().isEmpty) {
      throw ArgumentError('Title cannot be empty');
    }

    final existingTodo = await _repository.getById(id);
    if (existingTodo == null) {
      throw Exception('Todo with ID "$id" not found');
    }

    final updatedTodo = existingTodo.copyWith(
      title: title?.trim(),
      description: description?.trim(),
      deadline: deadline,
      isCompleted: isCompleted,
      iconType: iconType,
    );

    return await _repository.update(updatedTodo);
  }
}

File: apps/mobile/android/app/src/main/java/com/erluxman/productive/core/application/use_cases/UpdateTodoUseCase.kt
package com.erluxman.productive.core.application.use_cases

import com.erluxman.productive.core.domain.entities.TodoEntity
import com.erluxman.productive.core.domain.exceptions.ValidationException
import com.erluxman.productive.core.domain.repositories.TodoRepository

class UpdateTodoUseCase(
    private val repository: TodoRepository
) {
    suspend fun execute(todo: TodoEntity): Result<TodoEntity> {
        val trimmedTitle = todo.title.trim()
        
        if (trimmedTitle.isEmpty()) {
            return Result.failure(
                ValidationException("Title cannot be empty")
            )
        }
        
        if (trimmedTitle.length > 200) {
            return Result.failure(
                ValidationException("Title cannot exceed 200 characters")
            )
        }
        
        return repository.update(todo.copy(title = trimmedTitle))
    }
}

